`azurerm_private_endpoint` Try to add retry on creation

lonegunmanb opened the pull request:
Add retry on creation to fix #16182 .

lonegunmanb commented:
Acc tests:

=== RUN   TestAccPrivateEndpoint_basic
=== PAUSE TestAccPrivateEndpoint_basic
=== CONT  TestAccPrivateEndpoint_basic
--- PASS: TestAccPrivateEndpoint_basic (462.73s)
=== RUN   TestAccPrivateEndpoint_updateTag

=== PAUSE TestAccPrivateEndpoint_updateTag

=== CONT  TestAccPrivateEndpoint_updateTag
--- PASS: TestAccPrivateEndpoint_updateTag (634.01s)
=== RUN   TestAccPrivateEndpoint_requestMessage
=== PAUSE TestAccPrivateEndpoint_requestMessage
=== CONT  TestAccPrivateEndpoint_requestMessage
--- PASS: TestAccPrivateEndpoint_requestMessage (534.62s)
=== RUN   TestAccPrivateEndpoint_privateDnsZoneGroup
=== PAUSE TestAccPrivateEndpoint_privateDnsZoneGroup
=== CONT  TestAccPrivateEndpoint_privateDnsZoneGroup
--- PASS: TestAccPrivateEndpoint_privateDnsZoneGroup (519.38s)
=== RUN   TestAccPrivateEndpoint_privateDnsZoneRename
=== PAUSE TestAccPrivateEndpoint_privateDnsZoneRename
=== CONT  TestAccPrivateEndpoint_privateDnsZoneRename
--- PASS: TestAccPrivateEndpoint_privateDnsZoneRename (656.88s)
=== RUN   TestAccPrivateEndpoint_privateDnsZoneUpdate
=== PAUSE TestAccPrivateEndpoint_privateDnsZoneUpdate
=== CONT  TestAccPrivateEndpoint_privateDnsZoneUpdate
--- PASS: TestAccPrivateEndpoint_privateDnsZoneUpdate (790.60s)
=== RUN   TestAccPrivateEndpoint_privateDnsZoneRemove
=== PAUSE TestAccPrivateEndpoint_privateDnsZoneRemove
=== CONT  TestAccPrivateEndpoint_privateDnsZoneRemove
--- PASS: TestAccPrivateEndpoint_privateDnsZoneRemove (731.34s)
=== RUN   TestAccPrivateEndpoint_privateConnectionAlias
=== PAUSE TestAccPrivateEndpoint_privateConnectionAlias
=== CONT  TestAccPrivateEndpoint_privateConnectionAlias
--- PASS: TestAccPrivateEndpoint_privateConnectionAlias (465.02s)

tombuildsstuff commented on internal/services/network/private_endpoint_resource.go:
```diff
@@ -299,8 +300,17 @@ func resourcePrivateEndpointCreate(d *pluginsdk.ResourceData, meta interface{})
 	}
-	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
-		return fmt.Errorf("waiting for creation of Private Endpoint %q (Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
+	err = pluginsdk.Retry(d.Timeout(pluginsdk.TimeoutCreate), func() *resource.RetryError {
+		if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
+			return &resource.RetryError{
+				Err:       fmt.Errorf("waiting for creation of Private Endpoint %q (Resource Group %q): %+v", id.Name, id.ResourceGroup, err),
+				Retryable: strings.Contains(strings.ToLower(err.Error()), "retryable"),
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return err
```
unfortunately this won't do anything since it's only retrying on the poller/future.. since this is an API error this'd need to be fixed here - is there a Swagger issue tracking this upstream issue?

lonegunmanb commented:
Hi @tombuildsstuff , since the error message is:

>Message="A retryable error occurred." Details=[{"code":"ReferencedResourceNotProvisioned","message":"Cannot proceed with operation because resource /subscriptions/***/resourceGroups/ao-network/providers/Microsoft.Network/virtualNetworks/vnet-we/subnets/private-endpoints-snet used by resource /subscriptions/***/resourceGroups/acintbf3e-stamp-westeurope-rg/providers/Microsoft.Network/networkInterfaces/acintbf3e-westeurop-acr-pe.nic.8b3a01bb-91bd-4a50-b197-04119f1cb398 is not in Succeeded state. Resource is in Updating state and the last operation that updated/is updating the resource is PutSubnetOperation."}]

After searching for `Resource is in Updating state and the last operation that updated/is updating the resource is PutSubnetOperation.` I found this [issue](https://github.com/hashicorp/terraform-provider-azurerm/issues/2605), and a pr with retry on the poller fix the issue with the very similar error message. Do I need open an issue on the Swagger project? Thanks!

lonegunmanb commented:
Hello @tombuildsstuff , is there any chance to merge is pr since pr #3392 fixed very similar issue #2605 by retry, or should I open an issue on Azure service team? Thanks.

lonegunmanb commented:
Hello @tombuildsstuff , a kindly ping for this pr, is there anything I can do to get this pr been merged? Any advice will be appreciated.

tombuildsstuff commented:
@lonegunmanb this PR still polls the Read request - so this isn't going to change anything?

lonegunmanb commented:
@tombuildsstuff thanks for the guidance. Now I use retry on CreateOrUpate and polling, would you please give another review? Thanks!

Acc tests:

=== RUN   TestAccPrivateEndpoint_basic
=== PAUSE TestAccPrivateEndpoint_basic
=== CONT  TestAccPrivateEndpoint_basic
--- PASS: TestAccPrivateEndpoint_basic (549.82s)
=== RUN   TestAccPrivateEndpoint_updateTag
=== PAUSE TestAccPrivateEndpoint_updateTag
=== CONT  TestAccPrivateEndpoint_updateTag
--- PASS: TestAccPrivateEndpoint_updateTag (610.13s)
=== RUN   TestAccPrivateEndpoint_requestMessage
=== PAUSE TestAccPrivateEndpoint_requestMessage
=== CONT  TestAccPrivateEndpoint_requestMessage
--- PASS: TestAccPrivateEndpoint_requestMessage (520.74s)
=== RUN   TestAccPrivateEndpoint_privateDnsZoneGroup
=== PAUSE TestAccPrivateEndpoint_privateDnsZoneGroup
=== CONT  TestAccPrivateEndpoint_privateDnsZoneGroup
--- PASS: TestAccPrivateEndpoint_privateDnsZoneGroup (711.88s)

=== RUN   TestAccPrivateEndpoint_privateDnsZoneRename
=== PAUSE TestAccPrivateEndpoint_privateDnsZoneRename
=== CONT  TestAccPrivateEndpoint_privateDnsZoneRename
--- PASS: TestAccPrivateEndpoint_privateDnsZoneRename (718.42s)
=== RUN   TestAccPrivateEndpoint_privateDnsZoneUpdate
=== PAUSE TestAccPrivateEndpoint_privateDnsZoneUpdate
=== CONT  TestAccPrivateEndpoint_privateDnsZoneUpdate
--- PASS: TestAccPrivateEndpoint_privateDnsZoneUpdate (840.11s)
=== RUN   TestAccPrivateEndpoint_privateDnsZoneRemove
=== PAUSE TestAccPrivateEndpoint_privateDnsZoneRemove
=== CONT  TestAccPrivateEndpoint_privateDnsZoneRemove
--- PASS: TestAccPrivateEndpoint_privateDnsZoneRemove (855.77s)

=== RUN   TestAccPrivateEndpoint_privateConnectionAlias
=== PAUSE TestAccPrivateEndpoint_privateConnectionAlias

=== CONT  TestAccPrivateEndpoint_privateConnectionAlias
--- PASS: TestAccPrivateEndpoint_privateConnectionAlias (465.13s)

lonegunmanb commented:
Rebased with latest main branch, acc tests:

=== RUN   TestAccPrivateEndpoint_basic
=== PAUSE TestAccPrivateEndpoint_basic
=== CONT  TestAccPrivateEndpoint_basic
--- PASS: TestAccPrivateEndpoint_basic (378.15s)
=== RUN   TestAccPrivateEndpoint_updateTag

=== PAUSE TestAccPrivateEndpoint_updateTag

=== CONT  TestAccPrivateEndpoint_updateTag
--- PASS: TestAccPrivateEndpoint_updateTag (627.65s)
=== RUN   TestAccPrivateEndpoint_requestMessage
=== PAUSE TestAccPrivateEndpoint_requestMessage
=== CONT  TestAccPrivateEndpoint_requestMessage
--- PASS: TestAccPrivateEndpoint_requestMessage (478.36s)
=== RUN   TestAccPrivateEndpoint_privateDnsZoneGroup
=== PAUSE TestAccPrivateEndpoint_privateDnsZoneGroup
=== CONT  TestAccPrivateEndpoint_privateDnsZoneGroup
--- PASS: TestAccPrivateEndpoint_privateDnsZoneGroup (524.03s)
=== RUN   TestAccPrivateEndpoint_privateDnsZoneRename
=== PAUSE TestAccPrivateEndpoint_privateDnsZoneRename
=== CONT  TestAccPrivateEndpoint_privateDnsZoneRename
--- PASS: TestAccPrivateEndpoint_privateDnsZoneRename (677.02s)
=== RUN   TestAccPrivateEndpoint_privateDnsZoneUpdate
=== PAUSE TestAccPrivateEndpoint_privateDnsZoneUpdate
=== CONT  TestAccPrivateEndpoint_privateDnsZoneUpdate
--- PASS: TestAccPrivateEndpoint_privateDnsZoneUpdate (773.47s)
=== RUN   TestAccPrivateEndpoint_privateDnsZoneRemove
=== PAUSE TestAccPrivateEndpoint_privateDnsZoneRemove
=== CONT  TestAccPrivateEndpoint_privateDnsZoneRemove
--- PASS: TestAccPrivateEndpoint_privateDnsZoneRemove (787.14s)
=== RUN   TestAccPrivateEndpoint_privateConnectionAlias
=== PAUSE TestAccPrivateEndpoint_privateConnectionAlias

=== CONT  TestAccPrivateEndpoint_privateConnectionAlias
--- PASS: TestAccPrivateEndpoint_privateConnectionAlias (388.26s)
PASS

katbyte reviewed (changes requested):
Looking at the issue it seems like this is just a bandaid over the existing problem: provisioning multiple private endpoints causes them all to update the same dependant resource because terraform doesn't know about this dependancy? instead should we not be locking on the subnet id to prevent multiple changes from happening at the same time?

also could we add a test that provisions 5 or 6 to cover the use cases described in the issue?

katbyte commented on internal/services/network/private_endpoint_resource.go:
```diff
@@ -289,17 +290,32 @@ func resourcePrivateEndpointCreate(d *pluginsdk.ResourceData, meta interface{})
+					Err:       fmt.Errorf("creating Private Endpoint %q (Resource Group %q) due to missing 'group Id', ensure that the 'subresource_names' type is populated: %+v", id.Name, id.ResourceGroup, err),
+					Retryable: false,
+				}
+			} else {
```
test /wr

lonegunmanb commented:
Hello @katbyte , thanks for your advice, I've added lock on subnet id, and add new test to create multiple private endpoint instances, all the tests passed:

=== RUN   TestAccPrivateEndpoint_basic
=== PAUSE TestAccPrivateEndpoint_basic
=== CONT  TestAccPrivateEndpoint_basic
--- PASS: TestAccPrivateEndpoint_basic (366.77s)
=== RUN   TestAccPrivateEndpoint_updateTag

=== PAUSE TestAccPrivateEndpoint_updateTag

=== CONT  TestAccPrivateEndpoint_updateTag
--- PASS: TestAccPrivateEndpoint_updateTag (647.21s)
=== RUN   TestAccPrivateEndpoint_requestMessage
=== PAUSE TestAccPrivateEndpoint_requestMessage
=== CONT  TestAccPrivateEndpoint_requestMessage
--- PASS: TestAccPrivateEndpoint_requestMessage (410.05s)
=== RUN   TestAccPrivateEndpoint_privateDnsZoneGroup
=== PAUSE TestAccPrivateEndpoint_privateDnsZoneGroup
=== CONT  TestAccPrivateEndpoint_privateDnsZoneGroup
--- PASS: TestAccPrivateEndpoint_privateDnsZoneGroup (504.90s)
=== RUN   TestAccPrivateEndpoint_privateDnsZoneRename
=== PAUSE TestAccPrivateEndpoint_privateDnsZoneRename
=== CONT  TestAccPrivateEndpoint_privateDnsZoneRename
--- PASS: TestAccPrivateEndpoint_privateDnsZoneRename (597.69s)
=== RUN   TestAccPrivateEndpoint_privateDnsZoneUpdate
=== PAUSE TestAccPrivateEndpoint_privateDnsZoneUpdate
=== CONT  TestAccPrivateEndpoint_privateDnsZoneUpdate
--- PASS: TestAccPrivateEndpoint_privateDnsZoneUpdate (761.94s)
=== RUN   TestAccPrivateEndpoint_privateDnsZoneRemove
=== PAUSE TestAccPrivateEndpoint_privateDnsZoneRemove
=== CONT  TestAccPrivateEndpoint_privateDnsZoneRemove
--- PASS: TestAccPrivateEndpoint_privateDnsZoneRemove (765.35s)
=== RUN   TestAccPrivateEndpoint_privateConnectionAlias
=== PAUSE TestAccPrivateEndpoint_privateConnectionAlias
=== CONT  TestAccPrivateEndpoint_privateConnectionAlias
--- PASS: TestAccPrivateEndpoint_privateConnectionAlias (412.78s)
=== RUN   TestAccPrivateEndpoint_multipleInstances
=== PAUSE TestAccPrivateEndpoint_multipleInstances
=== CONT  TestAccPrivateEndpoint_multipleInstances
--- PASS: TestAccPrivateEndpoint_multipleInstances (601.73s)
PASS

Would you please give this pr another review? Thanks!

katbyte commented on internal/services/network/private_endpoint_resource.go:
```diff
@@ -289,17 +290,35 @@ func resourcePrivateEndpointCreate(d *pluginsdk.ResourceData, meta interface{})
-	future, err := client.CreateOrUpdate(ctx, id.ResourceGroup, id.Name, parameters)
-	if err != nil {
-		if strings.EqualFold(err.Error(), "is missing required parameter 'group Id'") {
-			return fmt.Errorf("creating Private Endpoint %q (Resource Group %q) due to missing 'group Id', ensure that the 'subresource_names' type is populated: %+v", id.Name, id.ResourceGroup, err)
-		} else {
-			return fmt.Errorf("creating Private Endpoint %q (Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
+	err = pluginsdk.Retry(d.Timeout(pluginsdk.TimeoutCreate), func() *resource.RetryError {
```
now that we are locking can we revert this and remove the retry? as it shouldn't be required

lonegunmanb replied:
Copy that.

katbyte reviewed (changes requested):
we also have a  test failure:
```
------- Stdout: -------
=== RUN   TestAccPrivateEndpoint_privateConnectionAlias
=== PAUSE TestAccPrivateEndpoint_privateConnectionAlias
=== CONT  TestAccPrivateEndpoint_privateConnectionAlias
    testcase.go:110: Step 1/2 error: Error running apply: exit status 1
        
        Error: creating Private Endpoint "acctest-privatelink-220809073512394901" (Resource Group "acctestRG-privatelink-220809073512394901"): network.PrivateEndpointsClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="InvalidPrivateLinkServiceId" Message="PrivateLinkServiceId Invalid private link service id: {0} passed in private link service connection: {1} passed in input is invalid" Details=[]
        
          with azurerm_private_endpoint.test,
          on terraform_plugin_test.tf line 80, in resource "azurerm_private_endpoint" "test":
          80: resource "azurerm_private_endpoint" "test" {
        
--- FAIL: TestAccPrivateEndpoint_privateConnectionAlias (213.61s)
FAIL
```

lonegunmanb commented:
Hi @katbyte , I believe it's a concurrent related issue, if we run this test alone then it will pass. I'll try to fix it.

lonegunmanb commented:
Hi @katbyte , thanks for reviewing this pr. I've tried to remove the retry, but to fix the failed test I found I have to add this retry back.

My theory is, the test failure is due to some performance and cache issue on the service side. When we execute this test alone it'll pass, but if we execute all private endpoint tests together, the service side will complain that the private link service we've passed is invalid.

In this test we used private link service's alias instead of id, so I guess when a private link service has been created, if we query its id immediately it will return an empty result occasionally, depends on how busy the service side were. I think it's a retriable error. I added a retry on this error message then the tests passed.

Since I've modified the update function, I also add an update in the test.

=== RUN   TestAccPrivateEndpoint_basic
=== PAUSE TestAccPrivateEndpoint_basic
=== CONT  TestAccPrivateEndpoint_basic
--- PASS: TestAccPrivateEndpoint_basic (378.77s)
=== RUN   TestAccPrivateEndpoint_updateTag
=== PAUSE TestAccPrivateEndpoint_updateTag
=== CONT  TestAccPrivateEndpoint_updateTag
--- PASS: TestAccPrivateEndpoint_updateTag (667.90s)
=== RUN   TestAccPrivateEndpoint_requestMessage
=== PAUSE TestAccPrivateEndpoint_requestMessage
=== CONT  TestAccPrivateEndpoint_requestMessage
--- PASS: TestAccPrivateEndpoint_requestMessage (507.50s)
=== RUN   TestAccPrivateEndpoint_privateDnsZoneGroup
=== PAUSE TestAccPrivateEndpoint_privateDnsZoneGroup
=== CONT  TestAccPrivateEndpoint_privateDnsZoneGroup
--- PASS: TestAccPrivateEndpoint_privateDnsZoneGroup (456.53s)
=== RUN   TestAccPrivateEndpoint_privateDnsZoneRename
=== PAUSE TestAccPrivateEndpoint_privateDnsZoneRename
=== CONT  TestAccPrivateEndpoint_privateDnsZoneRename
--- PASS: TestAccPrivateEndpoint_privateDnsZoneRename (677.53s)
=== RUN   TestAccPrivateEndpoint_privateDnsZoneUpdate
=== PAUSE TestAccPrivateEndpoint_privateDnsZoneUpdate
=== CONT  TestAccPrivateEndpoint_privateDnsZoneUpdate

--- PASS: TestAccPrivateEndpoint_privateDnsZoneUpdate (780.76s)
=== RUN   TestAccPrivateEndpoint_privateDnsZoneRemove
=== PAUSE TestAccPrivateEndpoint_privateDnsZoneRemove
=== CONT  TestAccPrivateEndpoint_privateDnsZoneRemove
--- PASS: TestAccPrivateEndpoint_privateDnsZoneRemove (805.95s)
=== RUN   TestAccPrivateEndpoint_privateConnectionAlias
=== PAUSE TestAccPrivateEndpoint_privateConnectionAlias
=== CONT  TestAccPrivateEndpoint_privateConnectionAlias
--- PASS: TestAccPrivateEndpoint_privateConnectionAlias (399.50s)
=== RUN   TestAccPrivateEndpoint_updateToPrivateConnectionAlias
=== PAUSE TestAccPrivateEndpoint_updateToPrivateConnectionAlias
=== CONT  TestAccPrivateEndpoint_updateToPrivateConnectionAlias
--- PASS: TestAccPrivateEndpoint_updateToPrivateConnectionAlias (562.12s)
=== RUN   TestAccPrivateEndpoint_multipleInstances
=== PAUSE TestAccPrivateEndpoint_multipleInstances
=== CONT  TestAccPrivateEndpoint_multipleInstances
--- PASS: TestAccPrivateEndpoint_multipleInstances (638.91s)
PASS

katbyte reviewed (changes requested):
that sounds like a bug on the API side? could we open an issue on the rest specs and link to it where we are doing the retry?

lonegunmanb commented:
> that sounds like a bug on the API side? could we open an issue on the rest specs and link to it where we are doing the retry?

Sure thing @katbyte , I'll open a ticket on this issue to the service team, meanwhile as this issue occurred on concurrent scenario, and it's not quite relevant to the issue we're trying to fix by this pr, could we accept this pr for now to solve user's issue, and when the service team fix this bug I'll open a new pr to remove this retry? Anyway, I'll open a ticket now. Thanks for your review.

katbyte commented on internal/services/network/private_endpoint_resource.go:
```diff
@@ -282,24 +283,61 @@ func resourcePrivateEndpointCreate(d *pluginsdk.ResourceData, meta interface{})
-			return fmt.Errorf("creating Private Endpoint %q (Resource Group %q) due to missing 'group Id', ensure that the 'subresource_names' type is populated: %+v", id.Name, id.ResourceGroup, err)
-		} else {
-			return fmt.Errorf("creating Private Endpoint %q (Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
+	err = pluginsdk.Retry(d.Timeout(pluginsdk.TimeoutCreate), func() *resource.RetryError {
```
```suggestion
	err = pluginsdk.Retry(d.Timeout(pluginsdk.TimeoutCreate), func() *resource.RetryError { // this is required due to a bug in the API https://github.com/Azure/azure-rest-api-specs/issues/20289
```

katbyte reviewed (approved):
Nope this is fine as long as there is an open issue describing what the issue is on the rest specs and its linked to in the code where the workaround exists. adding that now and it'll LGTM 🌩️
